# Import Necessary Libraries

In [147]:
import pandas as pd
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

%matplotlib inline

# Loading Data

In [148]:
df = pd.read_csv(r'../data/v1_EDA.csv')
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [149]:
df.shape

(13320, 9)

### Let's Check The duplicate values

In [150]:
duplicate_values = df.duplicated().sum()
print(f"Total Duplicates: {duplicate_values}")

Total Duplicates: 529


In [151]:
# Let's drop the duplicate rows
df = df.drop_duplicates()

In [152]:
# lets check again for duplicates
duplicate_values = df.duplicated().sum()
print(f"Total Duplicates after dropping: {duplicate_values} and the shape: {df.shape}")

Total Duplicates after dropping: 0 and the shape: (12791, 9)


In [153]:
13320 - 529

12791

### Let's Check The Missing values

In [154]:
print(df.isnull().sum())
print("\n")
print(df.shape)

area_type          0
availability       0
location           1
size              16
society         5328
total_sqft         0
bath              73
balcony          605
price              0
dtype: int64


(12791, 9)


In [155]:
# Check if rows missing 'size' are also missing 'society'
df[df['size'].isnull() & df['society'].isnull()]

area_type          availability      location size society total_sqft  \
2264   Plot  Area  Immediate Possession  Banashankari  NaN     NaN       2400   
7680   Plot  Area  Immediate Possession  Kasavanhalli  NaN     NaN       5000   
10634  Plot  Area  Immediate Possession    Rajankunte  NaN     NaN       1575   

       bath  balcony   price  
2264    NaN      NaN  460.00  
7680    NaN      NaN  400.00  
10634   NaN      NaN   31.11

so if we drop size column then society columns 3 rows will also be dropped.

In [156]:
# Check if rows missing 'size' are also missing 'bath'
df[df['size'].isnull() & df['bath'].isnull()]

area_type          availability        location size  society  \
579    Plot  Area  Immediate Possession  Sarjapur  Road  NaN  Asiss B   
1775   Plot  Area  Immediate Possession        IVC Road  NaN  Orana N   
2264   Plot  Area  Immediate Possession    Banashankari  NaN      NaN   
2809   Plot  Area  Immediate Possession  Sarjapur  Road  NaN  AsdiaAr   
2862   Plot  Area  Immediate Possession     Devanahalli  NaN  Ajleyor   
5333   Plot  Area  Immediate Possession     Devanahalli  NaN  Emngs S   
6423   Plot  Area  Immediate Possession      Whitefield  NaN  SRniaGa   
6636   Plot  Area  Immediate Possession          Jigani  NaN  S2enste   
6719   Plot  Area  Immediate Possession         Hoskote  NaN  SJowsn    
7680   Plot  Area  Immediate Possession    Kasavanhalli  NaN      NaN   
8306   Plot  Area  Immediate Possession          Anekal  NaN  Vaensnd   
8565   Plot  Area  Immediate Possession      Carmelaram  NaN  Ocezes    
8703   Plot  Area  Immediate Possession     Mysore Road  NaN  Naoden    
10634  Plot  Area  Immediate Possession      Rajankunte  NaN      NaN   
11019  Plot  Area  Immediate Possession       Yelahanka  NaN  Saaveha   
11569  Plot  Area  Immediate Possession      Hosur Road  NaN  AVeldun   

        total_sqft  bath  balcony    price  
579    1200 - 2400   NaN      NaN   34.185  
1775   2000 - 5634   NaN      NaN  124.000  
2264          2400   NaN      NaN  460.000  
2809   1200 - 2400   NaN      NaN   28.785  
2862   1500 - 2400   NaN      NaN   46.800  
5333   2100 - 5405   NaN      NaN  177.115  
6423          2324   NaN      NaN   26.730  
6636          1500   NaN      NaN   25.490  
6719    800 - 2660   NaN      NaN   28.545  
7680          5000   NaN      NaN  400.000  
8306          1453   NaN      NaN   16.500  
8565          2000   NaN      NaN  120.000  
8703   1200 - 2400   NaN      NaN   42.300  
10634         1575   NaN      NaN   31.110  
11019  1200 - 1800   NaN      NaN   12.750  
11569         1350   NaN      NaN    8.440

so if we drop size column then bath column 16 rows will also be dropped.

In [157]:
# Check if rows missing 'size' are also missing 'balcony'
df[df['size'].isnull() & df['balcony'].isnull()]

area_type          availability        location size  society  \
579    Plot  Area  Immediate Possession  Sarjapur  Road  NaN  Asiss B   
1775   Plot  Area  Immediate Possession        IVC Road  NaN  Orana N   
2264   Plot  Area  Immediate Possession    Banashankari  NaN      NaN   
2809   Plot  Area  Immediate Possession  Sarjapur  Road  NaN  AsdiaAr   
2862   Plot  Area  Immediate Possession     Devanahalli  NaN  Ajleyor   
5333   Plot  Area  Immediate Possession     Devanahalli  NaN  Emngs S   
6423   Plot  Area  Immediate Possession      Whitefield  NaN  SRniaGa   
6636   Plot  Area  Immediate Possession          Jigani  NaN  S2enste   
6719   Plot  Area  Immediate Possession         Hoskote  NaN  SJowsn    
7680   Plot  Area  Immediate Possession    Kasavanhalli  NaN      NaN   
8306   Plot  Area  Immediate Possession          Anekal  NaN  Vaensnd   
8565   Plot  Area  Immediate Possession      Carmelaram  NaN  Ocezes    
8703   Plot  Area  Immediate Possession     Mysore Road  NaN  Naoden    
10634  Plot  Area  Immediate Possession      Rajankunte  NaN      NaN   
11019  Plot  Area  Immediate Possession       Yelahanka  NaN  Saaveha   
11569  Plot  Area  Immediate Possession      Hosur Road  NaN  AVeldun   

        total_sqft  bath  balcony    price  
579    1200 - 2400   NaN      NaN   34.185  
1775   2000 - 5634   NaN      NaN  124.000  
2264          2400   NaN      NaN  460.000  
2809   1200 - 2400   NaN      NaN   28.785  
2862   1500 - 2400   NaN      NaN   46.800  
5333   2100 - 5405   NaN      NaN  177.115  
6423          2324   NaN      NaN   26.730  
6636          1500   NaN      NaN   25.490  
6719    800 - 2660   NaN      NaN   28.545  
7680          5000   NaN      NaN  400.000  
8306          1453   NaN      NaN   16.500  
8565          2000   NaN      NaN  120.000  
8703   1200 - 2400   NaN      NaN   42.300  
10634         1575   NaN      NaN   31.110  
11019  1200 - 1800   NaN      NaN   12.750  
11569         1350   NaN      NaN    8.440

so if we drop size column then balcony column 16 rows will also be dropped.

In [158]:
# Drop rows where 'size' or 'location' is missing
df = df.dropna(subset=['size', 'location'])
print(f"after dropping missing values: {df.shape}")

after dropping missing values: (12774, 9)


In [159]:
# Let's check again for missing values and shape
print(df.isnull().sum())
print("\n")
print(df.shape)

area_type          0
availability       0
location           0
size               0
society         5325
total_sqft         0
bath              57
balcony          589
price              0
dtype: int64


(12774, 9)


In [160]:
print(f"===> society was: {5328}\n now society is: 5328 - 3 = {5325}")
print(f"===> bath was: {73}\n now bath is: 73 - 16 = {57}")
print(f"===> balcony was: {605}\n now balcony is: 605 - 16 = {589}")

===> society was: 5328
 now society is: 5328 - 3 = 5325
===> bath was: 73
 now bath is: 73 - 16 = 57
===> balcony was: 605
 now balcony is: 605 - 16 = 589


In [161]:
12791 - 17

12774

In [162]:
# Let's Fill 'bath' with median
df['bath'] = df['bath'].fillna(df['bath'].median())
print(df.shape)

(12774, 9)


In [163]:
# Let's check again after filling bath missing values
print(df.isnull().sum())
print("\n")
print(df.shape)

area_type          0
availability       0
location           0
size               0
society         5325
total_sqft         0
bath               0
balcony          589
price              0
dtype: int64


(12774, 9)


In [164]:
# Let's Create a new columns based on size columns
df['bhk'] = df['size'].apply(lambda x: int(x.split(' ')[0]))
print(df.shape)
df.head()

(12774, 10)


area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  bhk  
0  Coomee        1056   2.0      1.0   39.07    2  
1  Theanmp       2600   5.0      3.0  120.00    4  
2      NaN       1440   2.0      3.0   62.00    3  
3  Soiewre       1521   3.0      1.0   95.00    3  
4      NaN       1200   2.0      1.0   51.00    2

In [165]:
#Drop the original 'size' column
df = df.drop(['size'], axis=1)
print(df.shape)
df.head()

(12774, 9)


area_type   availability                  location  society  \
0  Super built-up  Area         19-Dec  Electronic City Phase II  Coomee    
1            Plot  Area  Ready To Move          Chikka Tirupathi  Theanmp   
2        Built-up  Area  Ready To Move               Uttarahalli      NaN   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli  Soiewre   
4  Super built-up  Area  Ready To Move                  Kothanur      NaN   

  total_sqft  bath  balcony   price  bhk  
0       1056   2.0      1.0   39.07    2  
1       2600   5.0      3.0  120.00    4  
2       1440   2.0      3.0   62.00    3  
3       1521   3.0      1.0   95.00    3  
4       1200   2.0      1.0   51.00    2

In [166]:
# Let's check again after dropping size column
print(df.isnull().sum())
print("\n")
print(df.shape)

area_type          0
availability       0
location           0
society         5325
total_sqft         0
bath               0
balcony          589
price              0
bhk                0
dtype: int64


(12774, 9)


In [167]:
df['bhk'].unique()

array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18])

### Let's see the outlier

In [168]:
df[df.bhk>20]

area_type   availability                   location society  \
1718  Super built-up  Area  Ready To Move  2Electronic City Phase II     NaN   
4684            Plot  Area  Ready To Move                Munnekollal     NaN   

     total_sqft  bath  balcony  price  bhk  
1718       8000  27.0      0.0  230.0   27  
4684       2400  40.0      0.0  660.0   43

27 and 43-BHK house is extremely rare. </br>
these are data entry errors (like: someone entered the total square footage in the BHK column) </br>
If we have a 43-BHK house that is only 2,400 sqft, it's an impossible outlier that will confuse our model.

43 bedrooms but only 2400 sqft. Mathematically, that means each room is about 55 sqft (including bathrooms and hallways), which is impossible for a residential home.

Usually, a room needs at least 300 sqft. Anything significantly below that is likely a data entry error.

In [169]:
# Let's check total_sqft
df.total_sqft.unique()

<StringArray>
[       '1056',        '2600',        '1440',        '1521',        '1200',
        '1170',        '2732',        '3300',        '1310',        '1020',
 ...
        '3124',        '9200',         '613',         '250',        '2395',
 '1020 - 1130',        '2758', '1133 - 1384',         '774',        '4689']
Length: 2110, dtype: str

Our unique() check reveals that total_sqft is a StringArray containing values like '1020 - 1130'. Since a machine learning model only understands single numbers, we need to convert these ranges to their average.

In [170]:
# Let's check total_sqft is float or not
def is_float(x):
    try:
        float(x)
    except:
        return False
    return True

In [171]:
df[~df['total_sqft'].apply(is_float)].head(10)

area_type   availability            location  society  \
30   Super built-up  Area         19-Dec           Yelahanka  LedorSa   
56         Built-up  Area         20-Feb         Devanahalli  BrereAt   
81         Built-up  Area         18-Oct         Hennur Road  Gollela   
122  Super built-up  Area         18-Mar              Hebbal  SNontle   
137  Super built-up  Area         19-Mar  8th Phase JP Nagar  Vaarech   
165  Super built-up  Area         18-Dec            Sarjapur  Kinuerg   
188  Super built-up  Area  Ready To Move            KR Puram  MCvarar   
224  Super built-up  Area         19-Dec         Devanahalli  Jurdsig   
410  Super built-up  Area  Ready To Move             Kengeri      NaN   
549  Super built-up  Area         18-Sep         Hennur Road  Shxorm    

         total_sqft  bath  balcony    price  bhk  
30      2100 - 2850   4.0      0.0  186.000    4  
56      3010 - 3410   2.0      NaN  192.000    4  
81      2957 - 3450   2.0      NaN  224.500    4  
122     3067 - 8156   4.0      0.0  477.000    4  
137     1042 - 1105   2.0      0.0   54.005    2  
165     1145 - 1340   2.0      0.0   43.490    2  
188     1015 - 1540   2.0      0.0   56.800    2  
224     1520 - 1740   2.0      NaN   74.820    3  
410  34.46Sq. Meter   1.0      0.0   18.500    1  
549     1195 - 1440   2.0      0.0   63.770    2

In [172]:
# as we can see the there is massy data let's convert into number
def convert_sqft_to_num(x):
    tokens = x.split('-')
    if len(tokens) == 2:
        return (float(tokens[0]) + float(tokens[1]))/2
    try:
        return float(x)
    except:
        return None


In [173]:
convert_sqft_to_num('11225')

11225.0

In [174]:
# For getting average
convert_sqft_to_num('1145 - 1340')

1242.5

In [175]:
# For getting None
convert_sqft_to_num('34.46Sq. Meter')

In [176]:
df['total_sqft'] = df['total_sqft'].apply(convert_sqft_to_num)
df.head()

area_type   availability                  location  society  \
0  Super built-up  Area         19-Dec  Electronic City Phase II  Coomee    
1            Plot  Area  Ready To Move          Chikka Tirupathi  Theanmp   
2        Built-up  Area  Ready To Move               Uttarahalli      NaN   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli  Soiewre   
4  Super built-up  Area  Ready To Move                  Kothanur      NaN   

   total_sqft  bath  balcony   price  bhk  
0      1056.0   2.0      1.0   39.07    2  
1      2600.0   5.0      3.0  120.00    4  
2      1440.0   2.0      3.0   62.00    3  
3      1521.0   3.0      1.0   95.00    3  
4      1200.0   2.0      1.0   51.00    2

In [177]:
# Let's Check our 30 number index where total_sqft is 2100 - 2850 
df.iloc[30]

area_type       Super built-up  Area
availability                  19-Dec
location                   Yelahanka
society                      LedorSa
total_sqft                    2475.0
bath                             4.0
balcony                          0.0
price                          186.0
bhk                                4
Name: 30, dtype: object

In [178]:
# have a look it's cheange successfully 
(2100 + 2850)/2

2475.0

In [179]:
# Let's check after converting to number
print(df.isnull().sum())
print("\n")
print(df.shape)

area_type          0
availability       0
location           0
society         5325
total_sqft        46
bath               0
balcony          589
price              0
bhk                0
dtype: int64


(12774, 9)


Here we can see the **total_sqft** before it's 0 and after converted to number it's null values 46

In [180]:
# before Drop total_sqft let's check society and balcony will also be affected or not
total_missing_rows = df[df['total_sqft'].isnull() & df['society'].isnull()]
total_missing_rows

area_type   availability  \
410    Super built-up  Area  Ready To Move   
648          Built-up  Area  Ready To Move   
775          Built-up  Area  Ready To Move   
872    Super built-up  Area  Ready To Move   
1019             Plot  Area         18-Mar   
1086             Plot  Area         19-Mar   
1400   Super built-up  Area  Ready To Move   
3228         Built-up  Area  Ready To Move   
4113   Super built-up  Area         18-Jan   
4293         Built-up  Area  Ready To Move   
5453             Plot  Area  Ready To Move   
5668         Built-up  Area  Ready To Move   
6268         Built-up  Area  Ready To Move   
6303         Built-up  Area  Ready To Move   
6333             Plot  Area  Ready To Move   
6552             Plot  Area  Ready To Move   
6592         Built-up  Area  Ready To Move   
7001             Plot  Area  Ready To Move   
7248         Built-up  Area  Ready To Move   
7334             Plot  Area  Ready To Move   
7520   Super built-up  Area  Ready To Move   
7726             Plot  Area  Ready To Move   
8273         Built-up  Area  Ready To Move   
9423             Plot  Area  Ready To Move   
10488            Plot  Area  Ready To Move   
10491  Super built-up  Area  Ready To Move   
11320            Plot  Area  Ready To Move   
11615            Plot  Area  Ready To Move   
12186        Built-up  Area  Ready To Move   
12252        Built-up  Area  Ready To Move   

                                           location society  total_sqft  bath  \
410                                         Kengeri     NaN         NaN   1.0   
648                                         Arekere     NaN         NaN   9.0   
775                                    Basavanagara     NaN         NaN   2.0   
872                               Singapura Village     NaN         NaN   2.0   
1019                                 Marathi Layout     NaN         NaN   1.0   
1086                                     Narasapura     NaN         NaN   2.0   
1400                                     Chamrajpet     NaN         NaN   9.0   
3228                                Dodda Banaswadi     NaN         NaN   5.0   
4113                                     BTM Layout     NaN         NaN   3.0   
4293                              Bannerghatta Road     NaN         NaN   5.0   
5453                                         Kannur     NaN         NaN   6.0   
5668                                Judicial Layout     NaN         NaN   6.0   
6268                                       Chickpet     NaN         NaN   2.0   
6303                                       JP Nagar     NaN         NaN   3.0   
6333                                      Harohalli     NaN         NaN   2.0   
6552            5 Bedroom Farm House in Lakshmipura     NaN         NaN   6.0   
6592                                  Langford Town     NaN         NaN   4.0   
7001                               Thyagaraja Nagar     NaN         NaN   6.0   
7248                                       Hulimavu     NaN         NaN   2.0   
7334                                 Kanakpura Road     NaN         NaN   1.0   
7520                                  Doddaballapur     NaN         NaN   2.0   
7726                                 Kanakpura Road     NaN         NaN   1.0   
8273                                      V.V Puram     NaN         NaN   2.0   
9423                               Ramamurthy Nagar     NaN         NaN   1.0   
10488  2 Bedroom Furnished Farm House in Kolar Road     NaN         NaN   2.0   
10491                                      Rajapura     NaN         NaN   2.0   
11320                                 Arishinakunte     NaN         NaN   1.0   
11615                                         arudi     NaN         NaN   2.0   
12186                            7th Phase JP Nagar     NaN         NaN   2.0   
12252                                  Gowdanapalya     NaN         NaN   5.0   

       balcony   price  bhk  
410        0.0   18.50    1  
648        NaN  265

In [181]:
print(f"total_missing_rows shape: {total_missing_rows.shape[0]}")

total_missing_rows shape: 30


In [182]:
# before Drop total_sqft let's check society and balcony will also be affected or not
total_missing_rows = df[df['total_sqft'].isnull() & df['balcony'].isnull()]
total_missing_rows

area_type   availability           location  society  \
648         Built-up  Area  Ready To Move            Arekere      NaN   
872   Super built-up  Area  Ready To Move  Singapura Village      NaN   
2310  Super built-up  Area  Ready To Move         Whitefield  Haestt    
4293        Built-up  Area  Ready To Move  Bannerghatta Road      NaN   

      total_sqft  bath  balcony  price  bhk  
648          NaN   9.0      NaN  265.0    9  
872          NaN   2.0      NaN   45.0    2  
2310         NaN   4.0      NaN  270.0    4  
4293         NaN   5.0      NaN  260.0    4

In [183]:
print(f"total_missing_rows shape: {total_missing_rows.shape[0]}")

total_missing_rows shape: 4


In [184]:
# Drop those 46 new Nulls in total_sqft
df = df.dropna(subset=['total_sqft'])
print(df.shape)

(12728, 9)


In [185]:
12774 - 46

12728

In [186]:
# Let's check after dropping nulls in total_sqft   
print(df.isnull().sum())
print("\n")
print(df.shape)

area_type          0
availability       0
location           0
society         5295
total_sqft         0
bath               0
balcony          585
price              0
bhk                0
dtype: int64


(12728, 9)


<h3 style="color:white">Outlier Removal Using Business Logic</h3>

As a data scientist when you have a conversation with your business manager (who has expertise in real estate), he will tell you that normally square ft per bedroom is 300 (i.e. 2 bhk apartment is minimum 600 sqft. If you have for example 400 sqft apartment with 2 bhk than that seems suspicious and can be removed as an outlier. We will remove such outliers by keeping our minimum thresold per bhk to be 300 sqft

In [187]:
df[df.total_sqft/df.bhk<300].head()

area_type   availability             location society  total_sqft  bath  \
9   Plot  Area  Ready To Move         Gandhi Bazar     NaN      1020.0   6.0   
45  Plot  Area  Ready To Move           HSR Layout     NaN       600.0   9.0   
58  Plot  Area  Ready To Move        Murugeshpalya     NaN      1407.0   4.0   
68  Plot  Area  Ready To Move  Devarachikkanahalli     NaN      1350.0   7.0   
70  Plot  Area  Ready To Move          Double Road     NaN       500.0   3.0   

    balcony  price  bhk  
9       NaN  370.0    6  
45      NaN  200.0    8  
58      1.0  150.0    6  
68      0.0   85.0    8  
70      2.0  100.0    3

Check above data points. We have 6 bhk apartment with 1020 sqft. Another one is 8 bhk and total sqft is 600. These are clear data errors that can be removed safely

In [188]:
df[df.total_sqft/df.bhk<300].value_counts().sum()

np.int64(40)

In [189]:
df.shape

(12728, 9)

In [190]:
# Now we can Remove those outliers where total_sqft per bhk is less than 300 sqft
df[~(df.total_sqft / df.bhk < 300)]

area_type   availability                  location  society  \
0      Super built-up  Area         19-Dec  Electronic City Phase II  Coomee    
1                Plot  Area  Ready To Move          Chikka Tirupathi  Theanmp   
2            Built-up  Area  Ready To Move               Uttarahalli      NaN   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli  Soiewre   
4      Super built-up  Area  Ready To Move                  Kothanur      NaN   
...                     ...            ...                       ...      ...   
13314  Super built-up  Area  Ready To Move         Green Glen Layout  SoosePr   
13315        Built-up  Area  Ready To Move                Whitefield  ArsiaEx   
13316  Super built-up  Area  Ready To Move             Richards Town      NaN   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar  Mahla T   
13318  Super built-up  Area         18-Jun           Padmanabhanagar  SollyCl   

       total_sqft  bath  balcony   price  bhk  
0          1056.0   2.0      1.0   39.07    2  
1          2600.0   5.0      3.0  120.00    4  
2          1440.0   2.0      3.0   62.00    3  
3          1521.0   3.0      1.0   95.00    3  
4          1200.0   2.0      1.0   51.00    2  
...           ...   ...      ...     ...  ...  
13314      1715.0   3.0      3.0  112.00    3  
13315      3453.0   4.0      0.0  231.00    5  
13316      3600.0   5.0      NaN  400.00    4  
13317      1141.0   2.0      1.0   60.00    2  
13318      4689.0   4.0      1.0  488.00    4  

[11994 rows x 9 columns]

In [191]:
df[df.bhk>20]

area_type   availability                   location society  \
1718  Super built-up  Area  Ready To Move  2Electronic City Phase II     NaN   
4684            Plot  Area  Ready To Move                Munnekollal     NaN   

      total_sqft  bath  balcony  price  bhk  
1718      8000.0  27.0      0.0  230.0   27  
4684      2400.0  40.0      0.0  660.0   43

In [192]:
# Let's check finally
print(df.isnull().sum())
print("\n")
print(df.shape)

area_type          0
availability       0
location           0
society         5295
total_sqft         0
bath               0
balcony          585
price              0
bhk                0
dtype: int64


(12728, 9)


In [193]:
12728 - 11994

734

# Save Clean data

In [194]:
df.to_csv(r"../data/v2_Clean.csv", index=False)

"We noticed some houses had 43 bedrooms in just 2400 square feet. By creating a threshold of 300 sqft per room, we remove these 'dirty' data points. This ensures our model doesn't get confused by impossible scenarios, leading to much higher accuracy on real-world homes."

**Completed Steps**

**Duplicate Removal:** We identified and dropped **529** duplicate rows.

**Missing Value Handling (Initial):** We dropped rows where size and location were missing.

**Imputation:** We successfully filled missing **bath** values using the median.

**Feature Extraction:** We created the numerical **bhk** column from the **size** string and dropped the original size column.

**Structural Conversion:** We converted total_sqft from messy strings and ranges into clean floating-point numbers.

**Outlier Identification:** We identified illogical data points (e.g., 43 bedrooms in 2400 sqft).

### This is for ```src/components/b_cleaning.py```

In [195]:
import os
import sys
import pandas as pd
import numpy as np
import logging

logger = logging.getLogger(__name__)


class DataCleaning:
    def __init__(self):
        pass

    def convert_sqft_to_num(self, x):
        """Verified Logic: Converts ranges to average and handles units."""
        try:
            tokens = str(x).split('-')
            if len(tokens) == 2:
                return (float(tokens[0]) + float(tokens[1])) / 2
            return float(x)
        except Exception:
            return None

    def start_data_cleaning(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Finalized cleaning function based on verified notebook steps.
        """
        try:
            logger.info("Starting production data cleaning process...")
            df = df.copy()

            # 1. Drop Duplicates
            df = df.drop_duplicates()

            # 2. Handle Missing Values for critical features
            df = df.dropna(subset=['size', 'location'])

            # 3. Impute 'bath' with median
            df['bath'] = df['bath'].fillna(df['bath'].median())

            # 4. Create 'bhk' and drop 'size'
            df['bhk'] = df['size'].apply(lambda x: int(x.split(' ')[0]))
            df = df.drop(['size'], axis=1)

            # 5. Clean 'total_sqft'
            df['total_sqft'] = df['total_sqft'].apply(self.convert_sqft_to_num)
            
            # 6. Drop those 46 new Nulls in total_sqft
            df = df.dropna(subset=['total_sqft'])

            # 7. Business Logic Outlier Removal
            # Removing houses where sqft per room is less than 300
            df = df[~(df.total_sqft / df.bhk < 300)]
            
            logger.info(f"Cleaning complete. Resulting shape: {df.shape}")
            return df

        except Exception as e:
            logger.error("Error occurred in Data Cleaning component")
            raise 